core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Controller.scala

```diff
@@ -85,26 +96,29 @@ class Controller(
 
   def handleDirectInvocation: Receive = {
     case c: ControlInvocation =>
-      cp.processControlPayload(ChannelID(SELF, SELF, isControl = true), c)
+      // only client and self can send direction invocations
+      val source = if (sender == self) {
+        SELF
+      } else {
+        CLIENT
```

Comment:
try to align the message handling. we can have messages to optionally skip some steps.

Reply:
will do later since it can not be easily integrated into actor message now.